Source/FieldSolver/ImplicitSolvers/ThetaImplicitEM.cpp

```diff
@@ -0,0 +1,171 @@
+    if (m_nlsolver_type==NonlinearSolverType::Picard) {
+        amrex::Print() << "Nonlinear solver type:      Picard" << std::endl;
+    }
+    else if (m_nlsolver_type==NonlinearSolverType::Newton) {
+        amrex::Print() << "Nonlinear solver type:      Newton" << std::endl;
+    }
+    m_nlsolver->PrintParams();
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << std::endl;
+}
+
+void ThetaImplicitEM::OneStep ( const amrex::Real  a_time,
+                                const amrex::Real  a_dt,
+                                const int          a_step )
+{
+    amrex::ignore_unused(a_step);
+
+    // Fields have E^{n} and B^{n}
+    // Particles have p^{n} and x^{n}.
+
+    // Save the values at the start of the time step,
+    m_WarpX->SaveParticlesAtImplicitStepStart ( );
+
+    // Save the fields at the start of the step
+    m_Eold.Copy( m_WarpX->getMultiLevelField(FieldType::Efield_fp) );
+    m_E.Copy(m_Eold); // initial guess for E
+
+    const int num_levels = static_cast<int>(m_Bold.size());
+    for (int lev = 0; lev < num_levels; ++lev) {
+        for (int n=0; n<3; n++) {
+            const amrex::MultiFab& Bfp = m_WarpX->getField(FieldType::Bfield_fp,lev,n);
+            amrex::MultiFab& Bold = *m_Bold[lev][n];
+            amrex::MultiFab::Copy(Bold, Bfp, 0, 0, 1, Bold.nGrowVect());
+        }
+    }
+
+    const amrex::Real theta_time = a_time + m_theta*a_dt;
+
+    // Solve nonlinear system for E at t_{n+theta}
+    // Particles will be advanced to t_{n+1/2}
+    m_nlsolver->Solve( m_E, m_Eold, theta_time, a_dt );
+
+    // Update WarpX owned Efield_fp and Bfield_fp to t_{n+theta}
+    UpdateWarpXFields( m_E, theta_time, a_dt );
+
+    // Advance particles from time n+1/2 to time n+1
+    m_WarpX->FinishImplicitParticleUpdate();
+
+    // Advance E and B fields from time n+theta to time n+1
+    const amrex::Real new_time = a_time + a_dt;
+    FinishFieldUpdate( new_time );
+
+}
+
+void ThetaImplicitEM::ComputeRHS ( WarpXSolverVec&  a_RHS,
+                             const WarpXSolverVec&  a_E,
+                                   amrex::Real      a_time,
+                                   amrex::Real      a_dt,
+                                   int              a_nl_iter,
+                                   bool             a_from_jacobian )
+{
+    // update WarpX-owned Efield_fp and Bfield_fp using current state of E from
+    // the nonlinear solver at time n+theta
+    UpdateWarpXFields( a_E, a_time, a_dt );
+
+    // Self consistently update particle positions and velocities using the
+    // current state of the fields E and B. Deposit current density at time n+1/2.
+    m_WarpX->ImplicitPreRHSOp( a_time, a_dt, a_nl_iter, a_from_jacobian );
+
+    // RHS = cvac^2*m_theta*dt*( curl(B^{n+theta}) - mu0*J^{n+1/2} )
+    m_WarpX->ImplicitComputeRHSE(m_theta*a_dt, a_RHS);
```

Comment:
Just checking that I understand this correctly: the update of `E` (in `a_RHS`) actually uses `B^{n+theta}` from the previous iteration of the nonlinear solver, instead of that using the `B^{n+theta}` which is consistent with the new `E` at this nonlinear iteration. Is that correct?

Reply:
It uses B^{n+theta} computed with the current iteration value of E^{n+theta} (i.e., the passed in value a_E into the ComputeRHS function).

Reply:
![Uploading Picture1.png…]()

Reply:
![Uploading Picture1.png…]()